Introduce std.log

ifreund opened the pull request:
std.log provides 8 log levels and corresponding logging functions. It
allows the user to override the logging "backend" by defining root.log
and to override the default log level by defining root.log_level.

Logging functions accept a scope parameter which allows the implementer
of the logging "backend" to filter logging by library as well as level.

Using the standardized syslog [1] log levels ensures that std.log will
be flexible enough to work for as many use-cases as possible. If we were
to stick with only 3/4 log levels, std.log would be insufficient for
large and/or complex projects such as a kernel or display server.

[1]: https://tools.ietf.org/html/rfc5424#section-6.2.1
TODO:
- [x] Implement `std.log`
- [x] Deprecate `std.debug.warn`
- [ ] Update documentation
- [ ] Update standard library code using `std.debug.warn` (perhaps in a follow-up PR)

Will close #2586

dbandstra commented on lib/std/log.zig:
```diff
@@ -0,0 +1,74 @@
+    @setCold(true);
+    // On release small builds we don't log because that would require storing
+    // the error message.
+    if (!builtin.mode == .ReleaseSmall) log(.Fatal, format, args);
```
looks like a typo, should be `!=` ?

ifreund replied:
indeed, i seem to have placed the `!` before `builtin.mode` for some reason.

dbandstra commented on lib/std/log.zig:
```diff
@@ -0,0 +1,74 @@
+
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(message_level: Level, comptime format: []const u8, args: var) void {
```
Should the `message_level` arg be made comptime, to avoid code generation in release modes? Or is the zig compiler able to figure this out implicitly?

ifreund replied:
certainly couldn't hurt

SpexGuy replied:
As a real-world example, Unreal Engine uses macros to have both a comptime minimum log level and a runtime-controlled log level, so you can e.g. compile out verbose log statements but turn on/off info log statements without a recompile.

ifreund replied:
@SpexGuy I think the way #2586 is going is to keep the default logging only configurable at comptime. However, runtime control can be gained if desired by defining `root.log` which receives the log level any may implement comptime or runtime configuration.

Vexu commented on lib/std/log.zig:
```diff
@@ -0,0 +1,74 @@
+
+pub const Level = enum {
+    /// helpful information for debugging.
+    Info,
```
These should be snake_case #2101.

dbandstra commented on lib/std/log.zig:
```diff
@@ -0,0 +1,78 @@
+    // On release small builds we don't log because that would require storing
+    // the error message.
+    if (builtin.mode != .ReleaseSmall) log(.fatal, format, args);
+    unreachable;
```
I think this should be a panic. Release builds would optimize out the unreachable and continue executing

ifreund replied:
Good point, fixed

andrewrk commented on lib/std/log.zig:
```diff
@@ -0,0 +1,78 @@
+    @setCold(true);
+    // On release small builds we don't log because that would require storing
+    // the error message.
+    if (builtin.mode != .ReleaseSmall) log(.fatal, format, args);
```
This logic should probably go in the `log` function, so that it can be overridden by a custom `log` function, which may decide that logging is worth it, even for ReleaseSmall builds.

ifreund replied:
Makes sense, moved it!

data-man commented on lib/std/log.zig:
```diff
@@ -0,0 +1,81 @@
+/// log.default_level.
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(comptime message_level: Level, comptime format: []const u8, args: var) void {
```
```suggestion
pub fn log(comptime message_level: Level, comptime format: []const u8, args: var) void {
```

ifreund replied:
I believe that making this public would violate the "Only one obvious way to do things" bullet point of the zig zen. I prefer
```zig
std.log.warn("message", .{});
```
over 
```zig
std.log.log(.warn, "message", {});
```
personally so I'm inclined to keep this as is.

data-man replied:
`pub fn log` will support custom level if `Level` become non-exhaustive enum.

ifreund replied:
Frankly I'm not a fan of allowing custom log levels when using `std.log`. It is imo unnecessary complexity and will reduce compatibility between libraries and programs which is the whole point of a unified `std.log` interface.

data-man replied:
Lack of flexibility forces developers to reinvent their own wheels.

data-man commented:
Inspired by other libraries (in C++, Rust, Nim):
```zig
pub const Level = enum(u32) {
none, //or off, disabled, etc.
trace,
debug,
info,
notice,
warn,
err,
fatal,
_, //for custom level
};
```

daurnimator commented:
> Inspired by other libraries (in C++, Rust, Nim):
> 
> ```zig
> pub const Level = enum(u32) {
> none, //or off, disabled, etc.
> trace,
> debug,
> info,
> notice,
> warn,
> err,
> fatal,
> _, //for custom level
> };
> ```

These levels come from the [syslog protocol](https://tools.ietf.org/html/rfc5424#section-6.2.1) and is only a `u3`:

```
           Numerical         Severity
             Code

              0       Emergency: system is unusable
              1       Alert: action must be taken immediately
              2       Critical: critical conditions
              3       Error: error conditions
              4       Warning: warning conditions
              5       Notice: normal but significant condition
              6       Informational: informational messages
              7       Debug: debug-level messages

              Table 2. Syslog Message Severities
```

data-man commented:
Also would be nice supports the log function in structs if it's declared.

ifreund commented:
> Also would be nice supports the log function in structs if it's declared.

@data-man I'm not quite clear on what you're picturing here. Maybe you could give an example?

data-man commented:
@ifreund 
By analogy with `fmt.format`.

ifreund commented:
@data-man this already leverages `fmt.format` internally. Are you suggesting that there be a way to format things differently if they are passed to `log` as format arguments? I'm not sure I see the point in that.

data-man commented:
@ifreund 
> Are you suggesting that there be a way to format things differently if they are passed to log as format arguments?

No.
E.g. `File` hasn't `format`.
With my proposal you can add `log` fn to `File` and trace various things.

daurnimator commented on lib/std/log.zig:
```diff
@@ -0,0 +1,81 @@
+pub fn fatal(comptime format: []const u8, args: var) noreturn {
+    @setCold(true);
+    log(.fatal, format, args);
+    @panic("");
```
I don't think fatal logging should actually call `panic`: that should be up to the caller.

ifreund replied:
I'm conflicted on this, so I took my cue from @andrewrk's example in #2586. I'm leaning towards removing the panic and leaving it up to the caller myself though.

ifreund commented:
> I would suggest we also have a "component"/"facility" comptime argument.

I've been considering adding a `scope` argument in the form of a non-exhaustive enum to allow the implementer of `root.log` to filter, for example, `debug` level messages from a library being used. This seems like a pretty important feature and a non-exhaustive enum would be a good way to add it.

I don't plan on adding additional log levels as I think the current 4 are sufficient and keeping this count low will simplify filtering and make choices about what level to use easier.

ifreund commented:
I think removing the panic from `fatal` is the right thing to do, however once the panic is removed it is no longer different enough from `err` and so I removed the log level completely. I also added a new `silent` log level as it is a cleaner way to disable logging for release small build completely without special casing.

data-man commented:
@ifreund 
What is the visual difference between `info`, `warn` and `err` messages?
How can a user figure out what he sees?
E.g.:
```zig
info("Enter the Matrix", .{}); //Should be printed "[info] Enter the Matrix"?
warn("Enter the ", .{"Matrix"}); //Should be printed "[warning] Enter the Matrix"?
err("{} the {}", .{"Enter", "Matrix"}); //Should be printed "[error] Enter the Matrix"?
```

ifreund commented:
That's up to the user to implement. For example `root.log` might look like
```zig
pub fn log(comptime level: std.log.Level, comptime format: []const u8, args: var) void {
    const prefix = switch (level) {
        .info => "[INFO] ",
        .warn => "[WARN] ",
        .err => "[ERROR] ",
        .silent => return,
    };
    const held = std.debug.getStderrMutex().acquire();
    defer held.release();
    const stderr = std.debug.getStderrStream();
    nosuspend stderr.print(prefix ++ format, args) catch return;
}
```
I expect that most non-trivial programs will define `root.log` in a similar way. I don't think the standard library should force any particular formatting choice though.

If we opt to go with a more "batteries included" approach then something like this example could be made default, perhaps with some pretty colors if run from a tty.

data-man commented:
> That's up to the user to implement.

Sure.

> I don't think the standard library should force any particular formatting choice though.

If my previous example is inserted into the standard library, the user will see three identical messages.

daurnimator commented:
> I've been considering adding a `scope` argument in the form of a non-exhaustive enum to allow the implementer of `root.log` to filter, for example, `debug` level messages from a library being used. This seems like a pretty important feature and a non-exhaustive enum would be a good way to add it.

I don't see why you'd need a enum: just take a `comptime var` and let the user provide... anything.
Or if you wanted to reduce the options a little, accept a string. or an enum literal?

daurnimator commented:
> I think removing the panic from `fatal` is the right thing to do, however once the panic is removed it is no longer different enough from `err` and so I removed the log level completely. I also added a new `silent` log level as it is a cleaner way to disable logging for release small builds completely without special casing, and also allows the disabling logging for other build types.

I would advise following the syslog protocol levels I quoted earlier: they are pervasive across all logging systems. Sure it's hard to know when to use ALERT vs CRIT vs ERR.... but that's already a choice people make elsewhere.

ifreund commented:
I added a scope parameter in the form of an enum literal. I think that this gives the right amount of flexibility while still allowing for simple filtering with a swtich statement.

This now depends on #5203. You can also test by applying this simple patch
```diff
diff --git a/src/ir.cpp b/src/ir.cpp
index 6bd4e9b80..0bd44ee1f 100644
--- a/src/ir.cpp
+++ b/src/ir.cpp
@@ -25464,10 +25464,11 @@ static ZigType *type_info_to_type(IrAnalyze *ira, IrInst *source_instr, ZigTypeI
             ZigType *child_type = get_const_field_meta_type_optional(ira, source_instr->source_node, payload, "child", 0);
             return get_any_frame_type(ira->codegen, child_type);
         }
+        case ZigTypeIdEnumLiteral:
+            return ira->codegen->builtin_types.entry_enum_literal;
         case ZigTypeIdErrorSet:
         case ZigTypeIdEnum:
         case ZigTypeIdFnFrame:
-        case ZigTypeIdEnumLiteral:
             ir_add_error(ira, source_instr, buf_sprintf(
                 "TODO implement @Type for 'TypeInfo.%s': see https://github.com/ziglang/zig/issues/2907", type_id_name(tagTypeId)));
             return ira->codegen->invalid_inst_gen->value->type;
```

Still thinking about whether we want to follow the syslog protocol or keep things simple. The simplicity is appealing for smaller projects, though the added granularity of the syslog levels would certainly be useful at scale. I know that zig is intended to work well with massive code bases, so perhaps that is the way to go. It also has the advantage of being widely used an familiar already.

ifreund commented on lib/std/log.zig:
````diff
@@ -0,0 +1,131 @@
+//!     // Won't be printed as it gets filtered out by our log function
+//!     std.log.warn(.lib_that_logs_too_much, "Added 1 + 1\n", .{});
+//! }
+//! ```
````
Wasn't sure how example code should be formatted in doc comments, so I just defaulted to some markdown backticks

ifreund commented on lib/std/log.zig:
```diff
@@ -0,0 +1,131 @@
+    comptime format: []const u8,
+    args: var,
+) void {
+    if (@enumToInt(message_level) < @enumToInt(level))
+        return;
+    if (@hasDecl(root, "log")) {
+        root.log(message_level, scope, format, args);
```
I'm trying to decide if we should swap the order of these checks. With the current code one must set `root.log_level` to `.info` in order to implement runtime configurable log levels. However, this keeps `root.log` simpler in the case that comptime log level configuration is sufficient.

ifreund replied:
leaning towards keeping this as is. Setting `root.log_level` to `.info` makes sense if you want to handle filtering yourself at runtime.

ifreund commented:
After spending some time thinking on this, I've decided that @daurnimator is right and we should use the syslog logging levels.  I've implement them in this new commit and tweaked the rest of the code accordingly. The example the toplevel doc comment has also been updated. I will edit the PR description with the new commit message.

ifreund commented:
Deprecated `std.debug.warn` and added a `std.debug.print` function. I think it makes sense to have something like this for "printf debugging." Last thing this needs before merging is updated documentation if all the code changes look good.

ifreund commented:
Rebased to resolve conflicts.

joachimschmidt557 commented on lib/std/debug.zig:
```diff
@@ -52,15 +52,24 @@ pub const LineInfo = struct {
-/// Does not append a newline.
+/// Deprecated. Use `std.log` functions for logging or `std.debug.print` for
+/// "printf debugging".
 pub fn warn(comptime fmt: []const u8, args: var) void {
     const held = stderr_mutex.acquire();
     defer held.release();
     const stderr = io.getStdErr().writer();
     nosuspend stderr.print(fmt, args) catch return;
 }
```
```suggestion
pub const warn = print;
```

ifreund replied:
Good call

ifreund commented:
Note: this currently depends on `@Type()` for enum literals from #5203, though this could be avoided by using `@TypeOf(.foobar)` instead until that gets merged.
